File: Chapter - 2 -  Linked List/1. Remove Dups/RemoveDups.cpp
#include <iostream>
using namespace std;

struct Node{
	int data;
	Node* next;
	Node(int x){
		data = x;
		next = NULL;
	}
};

Node* pushToLast(Node* head, int data){
	Node* node = new Node(data);
	Node* temp = head;
	if(head == NULL){
		head = node;
	}else{
		while(temp && temp->next) temp = temp->next;
		temp->next = node;
	}
	return head;
}

Node* pushToFront(Node* head, int data){
	Node* node = new Node(data);
	Node* temp = head;
	if(temp == NULL) head = node;
	else{
		node->next = head;
		head = node;
	}
	return head;
}

int getLength(Node* head){
	Node* temp = head;
	int count = 0;
	while(temp) count++;
	return count;
}

void printList(Node* head){
	Node* temp = head;
	while(temp){
		cout << temp->data << " ";
		temp = temp->next;
	}
}

void RemoveDups(Node *head){
	bool vis[10000];
	for(int i = 0 ; i < 10000 ; i++) vis[i] = false;
	
	Node* temp = head;
	Node* ans = NULL;
	while(temp){
		if(vis[temp->data] == true){
			
		}else{
			vis[temp->data] = true;
			ans = pushToLast(ans, temp->data);
		}
		temp = temp->next;
	}
	cout << "New list after removal of duplicates is : ";
	printList(ans);
}

int main() {
	//Number of nodes in the linked list
	int n;
	cin >> n;
	Node* head = NULL;
	for(int i = 0 ; i < n ; i++){
		int x;
		cin >> x;
		head = pushToLast(head, x);
	}
	cout << "Old list was : ";
	printList(head);
	cout << endl;
	
	RemoveDups(head);
	
	return 0;
}

File: Chapter - 2 -  Linked List/1. Remove Dups/RemoveDups.java
import java.util.HashSet;
import java.util.Set;

class RemoveDups {
    static ListNode head = null;

    public static void removeDups(ListNode head) {
        Set<Integer> nums = new HashSet<>();
        ListNode curr = head;
        while (curr.next != null) {
            nums.add(curr.val);
            while (curr.next != null && nums.contains(curr.next.val)) {
                curr.next = curr.next.next;
                curr = curr.next;
            }
            curr = curr.next;
        }
    }

    public static void print(ListNode head) {
        ListNode current = head;
        while (current != null) {
            System.out.print(current.val + " ");
            current = current.next;
        }
        System.out.println();
    }

    public static void add(int data) {
        if (head == null) {
            head = new ListNode(data);
            return;
        }
        ListNode curr = head;
        while (curr.next != null) {
            curr = curr.next;
        }
        curr.next = new ListNode(data);
    }

    public static void main(String[] args) {
        add(3);
        add(5);
        add(3);
        add(67);
        add(5);
        add(90);
        add(43);

        print(head);
        removeDups(head);
        print(head);
    }

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}